chain/consensus/tspow/consensus.go

```diff
@@ -2,35 +2,232 @@ package tspow
 
 import (
 	"context"
+	"crypto/rand"
+	"fmt"
 
-	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/lib/async"
+	logging "github.com/ipfs/go-log/v2"
+	"golang.org/x/xerrors"
 
-	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/big"
+	lapi "github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/chain"
+	"github.com/filecoin-project/lotus/chain/beacon"
 	"github.com/filecoin-project/lotus/chain/consensus"
+	"github.com/filecoin-project/lotus/chain/stmgr"
 	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/lib/sigs"
+)
+
+const (
+	MaxDiffLookback = 70
+
+	// MaxHeightDrift is the epochs number to define blocks that will be rejected,
+	// if there are more than MaxHeightDrift epochs above the theoretical max height
+	// based on systime.
+	MaxHeightDrift = 5
 )
 
-var _ consensus.Consensus = (*TSPoW)(nil)
+var (
+	_   consensus.Consensus = &TSPoW{}
```

Comment:
Given Go's approach to duck typing of interfaces (which I dislike), I really like using this idiom to at least have compile-time type safety of interface implementations. This is nitpicking, but I only suggest having this line right above the struct definition with a comment like `// TSPoW implements consensus.Consensus`.

Reply:
Fixed